File: src/swift/fibonacciMemoization.swift
// Memoization dictionary to store computed Fibonacci numbers
var memo: [Int: Int] = [:]

// Function to calculate Fibonacci numbers using memoization
func fibonacci(_ n: Int) -> Int {
    if n <= 1 {
        return n
    }

    // Check if the result is already memoized
    if let cachedResult = memo[n] {
        return cachedResult
    }

    // Calculate Fibonacci recursively and store the result in memoization
    let result = fibonacci(n - 1) + fibonacci(n - 2)
    memo[n] = result

    return result
}

let index: Int = 15
print("Fibonacci (memoization):", fibonacci(index))

File: src/kotlin/FibonacciMemoization.kt
// Memoization dictionary to store computed Fibonacci numbers
val memo = mutableMapOf<Int, Long>()

// Function to calculate Fibonacci numbers using memoization
fun fibonacciMemoization(n: Int): Long {
    if (n <= 1) {
        return n.toLong()
    }

    // Check if the result is already memoized
    if (memo.containsKey(n)) {
        return memo[n]!!
    }

    // Calculate Fibonacci recursively and store the result in memoization
    val result = fibonacciMemoization(n - 1) + fibonacciMemoization(n - 2)
    memo[n] = result

    return result
}

fun main() {
    val index = 15
    println("Fibonacci (memoization) of $index is: ${fibonacciMemoization(index)}")
}
